# Day 4: Camp Cleanup

Space needs to be cleared before the last supplies can be unloaded from the ships, and so several Elves have been assigned the job of cleaning up sections of the camp. Every section has a unique ID number, and each Elf is assigned a range of section IDs.

However, as some of the Elves compare their section assignments with each other, they've noticed that many of the assignments overlap. To try to quickly find overlaps and reduce duplicated effort, the Elves pair up and make a big list of the section assignments for each pair (your puzzle input).

For example, consider the following list of section assignment pairs:

```
2-4,6-8
2-3,4-5
5-7,7-9
2-8,3-7
6-6,4-6
2-6,4-8
```

For the first few pairs, this list means:

- Within the first pair of Elves, the first Elf was assigned sections 2-4 (sections 2, 3, and 4), while the second Elf was assigned sections 6-8 (sections 6, 7, 8).
- The Elves in the second pair were each assigned two sections.
- The Elves in the third pair were each assigned three sections: one got sections 5, 6, and 7, while the other also got 7, plus 8 and 9.

This example list uses single-digit section IDs to make it easier to draw; your actual list might contain larger numbers. Visually, these pairs of section assignments look like this:

```
.234.....  2-4
.....678.  6-8

.23......  2-3
...45....  4-5

....567..  5-7
......789  7-9

.2345678.  2-8
..34567..  3-7

.....6...  6-6
...456...  4-6

.23456...  2-6
...45678.  4-8
```

Some of the pairs have noticed that one of their assignments fully contains the other. For example, 2-8 fully contains 3-7, and 6-6 is fully contained by 4-6. In pairs where one assignment fully contains the other, one Elf in the pair would be exclusively cleaning sections their partner will already be cleaning, so these seem like the most in need of reconsideration. In this example, there are 2 such pairs.

In how many assignment pairs does one range fully contain the other?

In [39]:
#@title Imports
import pandas as pd

In [40]:
#@title Import the data
df = pd.read_csv('input.txt', header=None)

In [41]:
df

0      1
0    54-59  17-62
1    20-93  57-92
2     6-54   7-54
3     3-99  59-98
4      5-8    5-8
..     ...    ...
995  59-89  68-89
996  84-86  57-84
997  24-94  54-93
998  41-63  40-63
999  87-90  89-90

[1000 rows x 2 columns]

In [42]:
"""
This function checks whether range_1 completely contains range_2.
For example 2-6,3-4 would return True but 2-6,5-7 would return False
"""
def completely_contains(range_1 : str, range_2 : str) -> bool:
    r1_start, r1_end = map(int, range_1.split('-'))
    r2_start, r2_end = map(int, range_2.split('-'))
    
    return r1_start <= r2_start and r1_end >= r2_end

def overlapping_range(range_1 : str, range_2 : str) -> bool:
    return int((completely_contains(range_1, range_2) or completely_contains(range_2, range_1)) == True)

In [43]:
#@title Test completely_contains and overlapping_range functions
import unittest

class TestOverlappingRange(unittest.TestCase):

    def test_cc_yes_within(self):
        self.assertEqual(completely_contains('2-6', '3-4'), True)

    def test_cc_yes_lower_boundary(self):
        self.assertEqual(completely_contains('2-6', '2-2'), True)

    def test_cc_yes_upper_boundary(self):
        self.assertEqual(completely_contains('2-6', '6-6'), True)

    def test_cc_no_too_low(self):
        self.assertEqual(completely_contains('2-6', '1-3'), False)

    def test_cc_no_too_high(self):
        self.assertEqual(completely_contains('2-6', '6-7'), False)

    def test_cc_no_total_miss_low(self):
        self.assertEqual(completely_contains('2-6', '1-1'), False)

    def test_cc_no_total_miss_high(self):
        self.assertEqual(completely_contains('2-6', '10-11'), False)

    def test_or_range_1(self):
        self.assertEqual(overlapping_range('2-6', '3-4'), 1)

    def test_or_range_2(self):
        self.assertEqual(overlapping_range('3-4', '2-6'), 1)

    def test_or_neither(self):
        self.assertEqual(overlapping_range('3-4', '10-11'), 0)

    

unittest.main(argv=[''], verbosity=2, exit=False)

test_contains_case_1 (__main__.TestContainsAtAll.test_contains_case_1) ... ok
test_contains_case_2 (__main__.TestContainsAtAll.test_contains_case_2) ... ok
test_contains_case_3 (__main__.TestContainsAtAll.test_contains_case_3) ... ok
test_contains_case_4 (__main__.TestContainsAtAll.test_contains_case_4) ... ok
test_contains_case_5 (__main__.TestContainsAtAll.test_contains_case_5) ... ok
test_contains_case_6 (__main__.TestContainsAtAll.test_contains_case_6) ... ok
test_contains_case_7 (__main__.TestContainsAtAll.test_contains_case_7) ... ok
test_cc_no_too_high (__main__.TestOverlappingRange.test_cc_no_too_high) ... ok
test_cc_no_too_low (__main__.TestOverlappingRange.test_cc_no_too_low) ... ok
test_cc_no_total_miss_high (__main__.TestOverlappingRange.test_cc_no_total_miss_high) ... ok
test_cc_no_total_miss_low (__main__.TestOverlappingRange.test_cc_no_total_miss_low) ... ok
test_cc_yes_lower_boundary (__main__.TestOverlappingRange.test_cc_yes_lower_boundary) ... ok
test_cc_yes_upper_bou

Case 3
Case 3
Case 1
Case 1
Case 1
Case 1
Case 1


In [44]:
#@title Clean and convert the data

# Add Column Names
df.columns = ['Elf 1', 'Elf 2']

# add score
df['Overlap'] = df.apply(lambda x: overlapping_range(x['Elf 1'], x['Elf 2']), axis=1)

In [45]:
#@title Print resulting data

df

Elf 1  Elf 2  Overlap
0    54-59  17-62        1
1    20-93  57-92        1
2     6-54   7-54        1
3     3-99  59-98        1
4      5-8    5-8        1
..     ...    ...      ...
995  59-89  68-89        1
996  84-86  57-84        0
997  24-94  54-93        1
998  41-63  40-63        1
999  87-90  89-90        1

[1000 rows x 3 columns]

In [46]:
#@title Calculate the answer
df['Overlap'].sum()

657

Your puzzle answer was 657.

The first half of this puzzle is complete! It provides one gold star: *

## Part Two

It seems like there is still quite a bit of duplicate work planned. Instead, the Elves would like to know the number of pairs that **overlap at all**.

In the above example, the first two pairs (2-4,6-8 and 2-3,4-5) don't overlap, while the remaining four pairs (5-7,7-9, 2-8,3-7, 6-6,4-6, and 2-6,4-8) do overlap:

- 5-7,7-9 overlaps in a single section, 7.
- 2-8,3-7 overlaps all of the sections 3 through 7.
- 6-6,4-6 overlaps in a single section, 6.
- 2-6,4-8 overlaps in sections 4, 5, and 6.

So, in this example, the number of overlapping assignment pairs is 4.

In how many assignment pairs do the ranges overlap?

In [47]:
"""
This function checks whether range_1 contains any of range_2.
For example 
2-4,6-8 and 2-3,4-5 don't overlap and return False
5-7,7-9 and 2-6,4-8  do overlap and return True
"""
def contains_at_all(range_1 : str, range_2 : str) -> bool:
    r1_start, r1_end = map(int, range_1.split('-'))
    r2_start, r2_end = map(int, range_2.split('-'))
    
    return r2_start <= r1_end and r2_end >= r1_start

In [48]:
#@title Test contains_at_all functions
import unittest

class TestContainsAtAll(unittest.TestCase):

    def test_contains_case_1(self):
        self.assertEqual(contains_at_all('2-4', '6-8'), False)

    def test_contains_case_2(self):
        self.assertEqual(contains_at_all('2-3', '4-5'), False)

    def test_contains_case_3(self):
        self.assertEqual(contains_at_all('5-7', '7-9'), True)

    def test_contains_case_4(self):
        self.assertEqual(contains_at_all('2-8', '3-7'), True)

    def test_contains_case_5(self):
        self.assertEqual(contains_at_all('6-6', '4-6'), True)

    def test_contains_case_6(self):
        self.assertEqual(contains_at_all('2-6', '4-8'), True)

    def test_contains_case_7(self):
        self.assertEqual(contains_at_all('1-2', '2-6'), True)

unittest.main(argv=[''], verbosity=2, exit=False)

test_contains_case_1 (__main__.TestContainsAtAll.test_contains_case_1) ... ok
test_contains_case_2 (__main__.TestContainsAtAll.test_contains_case_2) ... ok
test_contains_case_3 (__main__.TestContainsAtAll.test_contains_case_3) ... ok
test_contains_case_4 (__main__.TestContainsAtAll.test_contains_case_4) ... ok
test_contains_case_5 (__main__.TestContainsAtAll.test_contains_case_5) ... ok
test_contains_case_6 (__main__.TestContainsAtAll.test_contains_case_6) ... ok
test_contains_case_7 (__main__.TestContainsAtAll.test_contains_case_7) ... ok
test_cc_no_too_high (__main__.TestOverlappingRange.test_cc_no_too_high) ... ok
test_cc_no_too_low (__main__.TestOverlappingRange.test_cc_no_too_low) ... ok
test_cc_no_total_miss_high (__main__.TestOverlappingRange.test_cc_no_total_miss_high) ... ok
test_cc_no_total_miss_low (__main__.TestOverlappingRange.test_cc_no_total_miss_low) ... ok
test_cc_yes_lower_boundary (__main__.TestOverlappingRange.test_cc_yes_lower_boundary) ... ok
test_cc_yes_upper_bou

In [49]:
#@title Clean and convert the data

# add any overlap column
df['Any Overlap'] = df.apply(lambda x: int(contains_at_all(x['Elf 1'], x['Elf 2']) == True), axis=1)

In [50]:
#@title Print resulting data

df

Elf 1  Elf 2  Overlap  Any Overlap
0    54-59  17-62        1            1
1    20-93  57-92        1            1
2     6-54   7-54        1            1
3     3-99  59-98        1            1
4      5-8    5-8        1            1
..     ...    ...      ...          ...
995  59-89  68-89        1            1
996  84-86  57-84        0            1
997  24-94  54-93        1            1
998  41-63  40-63        1            1
999  87-90  89-90        1            1

[1000 rows x 4 columns]

In [51]:
#@title Calculate the answer
df['Any Overlap'].sum()

938